### DISCUSSION TITLE
Understanding how to implement plasticity with IsotropicPlasticityStressUpdate

### DISCUSSION THREAD

**Discussion by richmondodufisan** [idaholab/moose#26727] [Q&A Modules: Solid mechanics]
Hi, I would like to understand how to use this object to fully describe plastic deformation. As I understand plasticity theory, three things are needed to fully describe the system:

**-Yield Function**, i.e the function (typically stress-based i.e a function of the stress components) that calculates the yield surface, that is, where the material should begins to yield

**-Flow Rule**, i.e the function that explains how the material evolves beyond the yield surface

**-Hardening Law**: does the material harden isotropically (does the yield surface expand while staying centered about the origin?) or kinematically (does it remain constant in size but move with the changing yield stress?).

According to IsotropicPlasticityStressUpdate, it implements the radial return algorithm for a Von Mises yield function, but can be used in conjunction with more complex ones (and hopefully, custom yield functions). Looking at one of the example files, we have:

```
[Materials]
  [./elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 2.1e5
    poissons_ratio = 0.3
  [../]
  [./isotropic_plasticity]
    type = IsotropicPlasticityStressUpdate
    yield_stress = 50.0
    hardening_function = hf
  [../]
  [./radial_return_stress]
    type = ComputeMultipleInelasticStress
    tangent_operator = elastic
    inelastic_models = 'isotropic_plasticity'
  [../]
[]
```

I see that the elasticity tensor is defined, a single value for the yield stress is defined, and a hardening function "hf" is also defined.

**Q1: Where would I define my custom yield function to be compared with the yield stress?** As mentioned before, the default is the Von Mises stress. If I create a material file to calculate the custom yield surface based on the components of the elastic stress (which is initially calculated before doing a radial return), how do I let the model know to use that instead of the default Von Mises Stress?

**Q1b: What if the yield function cannot be easily expressed in the form sigma_yield - f(sigma_ii) = 0?** And instead is "implicit" i.e defined as f(sigma_yield, sigma_ii) = 0? For example, the Hoek-Brown Yield Criterion. Would that mean I can't use this object?

**Q2: how do I "extract" which stress I want to use?** If the model only used ComputeLinearElasticStress, then in a custom material file, I would "get" this stress with 

`_stress(getADMaterialPropertyByName<RankTwoTensor>("stress")),`

In this case where I have multiple stresses being calculated, what are the names of the stresses? For example, if I have both plasticity and viscoelasticity, what is the name of the "overall" stress (i.e the true stress experienced by the material) that I need to "get"? What about the name of the elastic stress? (So I can use it to check if the material is yielding).

**Q3: How does the model know which elastic stress to use (finite or small strain)?** In the ComputeMultipleInelasticStress, I see `tangent_operator = elastic`. Is the computation based on the strain defined? If I used, say, ADComputeGreenLagrangeStrain, would it automatically compute the PK2 stress?

**Q4: Where do I define the flow rule?** My guess is that the line `hardening_function = hf` is supposed to, but how? For an associative flow rule, it would be the derivative of the yield function with respect to stress. Where is that computed? What if I wanted to implement a custom, non-associative flow rule?

**Q5: Where do I define the hardening law?** That is, whether or not the yield function moves or expands in both compressive & tensile directions simultaneously?

**Comment by GiudGiud** [comment]
Hello

This material you are examining is a simple one with constants for the coefficients.

You need to look at the other materials in the module to fit your needs for more complex dependencies

I think the hardening laws are specified separately 

Guillaume

**Reply by richmondodufisan** [reply]
Even for this simple one, where is the Von Mises criterion defined? What if I wanted to use an even simpler one (i.e Tresca criterion)? And all the other questions still apply, i.e How is the stress "gotten" in the material file and how do I differentiate between each type of stress (elastic/trial or actual)? What does hf mean here?

```
  [./hf]
    type = PiecewiseLinear
    x = '0  0.001 0.003 0.023'
    y = '50 52    54    56'
  [../]
```

The flow rule is supposed to describe the evolution of plastic strains as a function of stress. What is this piecewise linear function describing? In the above we only have `tangent_operator = elastic` but there are different elastic stress/strain measures (finite, small strain, etc). Where/how are the hardening laws specified?

**Reply by GiudGiud** [reply]
>  where is the Von Mises criterion defined?

You can compute it with this for example
```
[AuxVariables]
  [./von_mises]
    #Dependent variable used to visualize the Von Mises stress
    order = CONSTANT
    family = MONOMIAL
  [../]
[]

[AuxKernels]
  [./von_mises_kernel]
    #Calculates the von mises stress and assigns it to von_mises
    type = RankTwoScalarAux
    variable = von_mises
    rank_two_tensor = stress
    execute_on = timestep_end
    scalar_type = VonMisesStress
  [../]
[]
```
Feel free to look through the tensor mechanics module for examples. 

> What if I wanted to use an even simpler one (i.e Tresca criterion)

You ll need to implement it

> How is the stress "gotten" in the material file

For output or for use in the material properties?

> and how do I differentiate between each type of stress (elastic/trial or actual)?

They are stored in different material properties.

> What does hf mean here?

It's just a name. You use that name to refer to the function in other places

> Where/how are the hardening laws specified?

In the Materials block. You should look through the examples in the tensor mechanics module

**Reply by richmondodufisan** [reply]
Thanks, I looked at a couple of examples and this makes sense. I have a couple more questions though:

-For the trial stress, in the documentation it is assumed linear elastic. From that elastic stress, the Von Mises stress is calculated, and then that is compared with the yield stress. What I wanted to know is how/where this trial stress is calculated? I am asking because I could potentially want to implement my own stress model (say a hyperelastic model) to combine with plasticity. So I wanted to understand how to integrate it with the IsotropicPlasticityStressUpdate.

-When in the plastic regime, the strains can now be divided into elastic and plastic strain. Is the strain object, say ComputeFiniteStrain "aware" of this? I just posted another question (https://github.com/idaholab/moose/discussions/27196), but I was asking where the `elastic_strain` is declared. How/where does it calculate and declare this, and does the plasticity object "override" the `total_strain` output? I guess my overall question is how this works, to make sure that when I "get" the `elastic_strain`, I am not getting any plastic part at all.

-Lastly, if I wanted to work without using this object (maybe if the model gets too complicated to integrate with this object) e.g if I used if/else statements to calculate the stress when in the plastic regime, how would I work with ComputeFiniteStrain to separate the plastic and elastic strains? I think all my questions really boil down to understanding how ComputeFiniteStrain works.

**Comment by jiangwen84** [comment]
> Thanks, I looked at a couple of examples and this makes sense. I have a couple more questions though:
> 
> -For the trial stress, in the documentation it is assumed linear elastic. From that elastic stress, the Von Mises stress is calculated, and then that is compared with the yield stress. What I wanted to know is how/where this trial stress is calculated? I am asking because I could potentially want to implement my own stress model (say a hyperelastic model) to combine with plasticity. So I wanted to understand how to integrate it with the IsotropicPlasticityStressUpdate.

See ComputeMultipleInelasticStress.C line 77
> 
> -When in the plastic regime, the strains can now be divided into elastic and plastic strain. Is the strain object, say ComputeFiniteStrain "aware" of this? I just posted another question (#27196), but I was asking where the `elastic_strain` is declared. How/where does it calculate and declare this, and does the plasticity object "override" the `total_strain` output? I guess my overall question is how this works, to make sure that when I "get" the `elastic_strain`, I am not getting any plastic part at all.
> 

ComputeFiniteStrain just computes the total strain/strain incremental from displacements. The separation of strain into elastic and plastic will be performed for the plasticity models called by ComputeMultipleInelasticStress. 

> -Lastly, if I wanted to work without using this object (maybe if the model gets too complicated to integrate with this object) e.g if I used if/else statements to calculate the stress when in the plastic regime, how would I work with ComputeFiniteStrain to separate the plastic and elastic strains? I think all my questions really boil down to understanding how ComputeFiniteStrain works.

As I explained, the ComputeFiniteStrain does not handle the plastic and elastic strains. You need to use ReturnMapping or other methods to get the plastic strain from the total strain. Usually this is a nonlinear equation, so we need newton iterations within return mapping.
